### PR TITLE
Apply Clippy lints regarding lazy evaluation and closures

### PR DESCRIPTION
**Pull request by Luracasmus**
# Objective

- Lazily evaluate [default](https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_or_default)~~/[or](https://rust-lang.github.io/rust-clippy/master/index.html#/or_fun_call)~~ values where it makes sense
  - ~~`unwrap_or(foo())` -> `unwrap_or_else(|| foo())`~~
  - `unwrap_or(Default::default())` -> `unwrap_or_default()`
  - etc.
- Avoid creating [redundant closures](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure), even for [method calls](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure_for_method_calls)
  - `map(|something| something.into())` -> `map(Into:into)`

## Solution

- Apply Clippy lints:
  - ~~[or_fun_call](https://rust-lang.github.io/rust-clippy/master/index.html#/or_fun_call)~~
  - [unwrap_or_default](https://rust-lang.github.io/rust-clippy/master/index.html#/unwrap_or_default)
  - [redundant_closure_for_method_calls](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure_for_method_calls) ([redundant closures](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure) is already enabled)

## Testing

- Tested on Windows 11 (`stable-x86_64-pc-windows-gnu`, 1.79.0)
  - Bevy compiles without errors or warnings and examples seem to work as intended
  - `cargo clippy` ✅
  - `cargo run -p ci -- compile` ✅

### CODE REVIEW

File: crates/bevy_dev_tools/src/ui_debug_overlay/mod.rs [hunk 1]
```diff
@@ -219,7 +219,7 @@ fn outline_roots(
         // We skip ui in other windows that are not the primary one
         if let Some(camera_entity) = maybe_target_camera
             .map(|target| target.0)
-            .or(cam.default_ui_camera.get())
+            .or_else(|| cam.default_ui_camera.get())
```
Comment (alice-i-cecile) [inline crates/bevy_dev_tools/src/ui_debug_overlay/mod.rs:222]: Not a fan of this change: which lint is this coming from?

File: crates/bevy_ecs/src/world/mod.rs [hunk 2]
```diff
@@ -3295,7 +3295,7 @@ mod tests {
         assert_eq!(world.entity(b2).get(), Some(&B(4)));
 
         let mut entities = world.iter_entities_mut().collect::<Vec<_>>();
-        entities.sort_by_key(|e| e.get::<A>().map(|a| a.0).or(e.get::<B>().map(|b| b.0)));
+        entities.sort_by_key(|e| e.get::<A>().map(|a| a.0).or_else(|| e.get::<B>().map(|b| b.0)));
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:3298]: Ditto on this change: I think this is less clear and introduces a closure.

File: crates/bevy_math/src/direction.rs [hunk 3]
```diff
@@ -158,7 +158,7 @@ impl Dir2 {
 
         direction
             .map(|dir| (Self(dir), length))
-            .ok_or(InvalidDirectionError::from_length(length))
+            .ok_or_else(|| InvalidDirectionError::from_length(length))
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/direction.rs:161]: I don't like these changes either :)

File: crates/bevy_reflect/src/path/access.rs [hunk 4]
```diff
@@ -55,7 +55,7 @@ impl<'a> Access<'a> {
         offset: Option<usize>,
     ) -> Result<&'r dyn Reflect, AccessError<'a>> {
         self.element_inner(base)
-            .and_then(|opt| opt.ok_or(AccessErrorKind::MissingField(base.reflect_kind())))
+            .and_then(|opt| opt.ok_or_else(|| AccessErrorKind::MissingField(base.reflect_kind())))
```
Comment (alice-i-cecile) [inline crates/bevy_reflect/src/path/access.rs:58]: More of the closure-inducing lint here.

File: crates/bevy_ui/src/focus.rs [hunk 5]
```diff
@@ -239,7 +239,7 @@ pub fn ui_focus_system(
             let camera_entity = node
                 .target_camera
                 .map(TargetCamera::entity)
-                .or(default_ui_camera.get())?;
+                .or_else(|| default_ui_camera.get())?;
```
Comment (alice-i-cecile) [inline crates/bevy_ui/src/focus.rs:242]: And more here.

File: crates/bevy_dev_tools/src/ui_debug_overlay/mod.rs (hunk 1 above)
Comment (Luracasmus) [reply crates/bevy_dev_tools/src/ui_debug_overlay/mod.rs:222]: All of the closure-inducing changes are from [or_fun_call](https://rust-lang.github.io/rust-clippy/master/index.html#/or_fun_call)

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
I dislike the effects of [`or_fun_call`](https://rust-lang.github.io/rust-clippy/master/#/or_fun), we should revert that change and not include it in our lints. We should manually check for places where it's actually saving us work though.

I like the others though, even if `Into::into` is a bit eh.

**Comment by alice-i-cecile** [comment]
Possibly: we should split that out into its own PR and review it there IMO.

**Comment by janhohenheim** [comment]
Maybe it's because I'm used to this lint for my own projects, but I find the code as presented in this PR more idiomatic / closer to how I would intuitively have written it in the first place. But yeah, I'm biased here :)

**Review by janhohenheim — APPROVED** [review]

**Comment by Luracasmus** [comment]
Are we still waiting on a review from @mockersf ? And if not, which solution do we go with? Including [or_fun_call](https://rust-lang.github.io/rust-clippy/master/index.html#/or_fun_call) or not (given the mixed feedback)?

**Comment by janhohenheim** [comment]
When in doubt, prefer @alice-i-cecile's feedback over mine, as she has to actually maintain the code in question.

**Comment by alice-i-cecile** [comment]
Cut or_fun_call from this PR please; I think it's worth doing selectively but not as a lint. The other changes look good and I'd like to unblock them.

**Review by alice-i-cecile — APPROVED** [review]
Good :) It looks like this needs to be formatted then I'll merge.

**Comment by Luracasmus** [comment]
> Good :) It looks like this needs to be formatted then I'll merge.

👍How do I format? Sry I'm not used to this workflow lol

**Comment by alice-i-cecile** [comment]
Run `cargo fmt --all` in your IDE :) The job with the red X on it will tell you more if you click "Details".

**Comment by Luracasmus** [comment]
> Run `cargo fmt --all` in your IDE :) The job with the red X on it will tell you more if you click "Details".

Ah I see, ty :)

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@Luracasmus I'll merge this once CI is passing.

**Pull request merged**